### PR TITLE
Plugin field mapping / priority of update flow / Salesforce sync push command

### PR DESCRIPTION
**Pull request by mqueme**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR changes how fields are mapped in plugins. Firstly it will show the fields you have saved on your plugins already, if you want to add another field mapping with your integration then you can do so by clicking on add field. you will be able to select what integration field will be mapped with a mautic field and also indicate the direction of the priority of the field. so when you sync it will say how fields should be overwritten.
[//]: # ( As applicable: )
#### Steps to test this PR:
1. First check that fields on all your previously configured plugins are still showing.
2. Select fields and check to see if they are updated in Mautic and if they are to be pushed to an integration then check to see that only the fields with an arrow pointing to the integration fields are updated in your integration with Mautic fields.
3. If you have a salesforce integration check that a new feature is added to push leads to integration and the previous one was renamed to triggered action (this one continues to work the same, as it will push a contact through an action in a form or campaign). 
4. To test the sync command select this new feature, save and run this command on CLI: 
`php app/console mautic:integration:fetchleads --time-interval="29 days" --env=dev --integration=Salesforce` it should run the command where it will either update or create leads/contacts in Salesforce

### CODE REVIEW

**Comment by escopecz** [comment]
I lose the fields in the contact field mapping tab. This is on staging branch:
screen shot 2017-02-28 at 16 59 10
Notice there is Last Name, Company, First Name, Email and several more fields bellow the fold.
If I apply this PR, I see this:
screen shot 2017-02-28 at 17 00 30
The ID field is mapped to the last name and other field mapping is lost. The Company mapping seems to stick even with the new changes.

**Comment by escopecz** [comment]
The fields are there, but
1. Is something happening when the arrow changes to animated circle?
2. Is there a reason why the left arrow background is red and right green?
3. The tooltip shows only on button background. It hides if I move cursor to the arrow icon.
4. Why are there some fields duplicated, once with __FirstName Lead__ and second with __FirstName Contact__?
screen shot 2017-03-01 at 10 22 23

**Comment by mqueme** [comment]
The duplicate are a salesforce thing. It's importing fields from lead and contact objects into mautic contacts

File: app/bundles/PluginBundle/Controller/PluginController.php [hunk 1]
```diff
@@ -205,14 +205,44 @@ public function configAction($name)
...
+                                        if (!empty($v) && strstr($f, 'm_') && isset($integrationField)) {
+                                            $mauticField                                         = $v;
+                                            $featureSettings['companyFields'][$integrationField] = $mauticField;
+                                        }
+                                    } else {
+                                        $featureSettings['update_mautic_company'][$integrationField] = (int) $v;
+                                    }
+                                }
+                            }
+                            $this->factory->getLogger()->addError(print_r($featureSettings, true));
```
Comment (escopecz) [inline app/bundles/PluginBundle/Controller/PluginController.php]: Isn't this a forgotten debug code?

File: plugins/MauticCrmBundle/Api/SalesforceApi.php [hunk 2]
```diff
@@ -240,31 +245,37 @@ public function getLeads($query, $object)
...
             case 'company':
             case 'Account':
-                $fields = array_keys(array_filter($fields['companyFields']));
+              //  $companyFieldsToUpdateInMautic = isset($fields['update_mautic_company']) ? array_keys($fields['update_mautic_company'], 0) : [];
+             //   $fields                        = array_keys(array_filter(array_diff_key($fields['companyFields'], $companyFieldsToUpdateInMautic)));
+              $fields = array_keys(array_filter($fields['companyFields']));
                 break;
             default:
+               // $fieldsToUpdateInMautic = isset($fields['update_mautic']) ? array_keys($fields['update_mautic'], 0) : [];
+               // $mixedFields            = array_filter(array_diff_key($fields['leadFields'], $fieldsToUpdateInMautic));
```
Comment (escopecz) [inline plugins/MauticCrmBundle/Api/SalesforceApi.php]: Maybe these commented lines should be removed?

File: app/bundles/PluginBundle/Views/FormTheme/Integration/_integration_details_featureSettings_leadFields_row.html.php [hunk 3]
```diff
@@ -16,24 +16,49 @@
...
+                <?php endif; ?>
+                    <?php ++$rowCount; ?>
+                    <div class="<?php if ($rowCount % $numberOfFields == 1) {
+    echo 'ml-lg';
+} ?> pl-xs pr-xs col-sm-<?php if ($rowCount % $numberOfFields == 2) {
+    echo '2 ml-xs';
+} else {
+    echo '4';
+} ?>">
+                        <?php echo $view['form']->row($child); ?>
```
Comment (escopecz) [inline app/bundles/PluginBundle/Views/FormTheme/Integration/_integration_details_featureSettings_leadFields_row.html.php]: I think it's recommended to use the `<?php if (...) : ?>...<?php endif; ?>` syntax in the HTML templates.

**Comment by npracht** [comment]
Hi @mqueme we're not done yet with Sugar. I hope we can push something friday or monday. We'll use the work you've done here for the priority :)

**Comment by escopecz** [comment]
1, 2, 3 fixed. 4 understood.

Right now the arrows are based on the radio button. It can be either one or the second option. Not both. It'd be great if both options could be checked. Another question is what happens to those fields which won't have any option checked. Will they be ignored? If so, maybe the arrows should be checked by default.

**Comment by mqueme** [comment]
The mautic direction is now default, I think it makes sense. and the direction is now only used when updating but when creating on either direction all the matched fields will be used.

**Comment by escopecz** [comment]
I get an error if I'm fetching SF leads:
```
$ php app/console mautic:integration:fetchleads --integration=Salesforce
Command to fetch contacts from integration
Fetch contacts command is starting
Number of contacts fetched: 2

Fetching companies
Fetch companies command is starting
Number of companies fetched: 0

Updating/creating leads from Matic to Salesforce
PHP Notice:  Undefined variable: body in plugins/MauticCrmBundle/Integration/SalesforceIntegration.php on line 959
PHP Notice:  Undefined variable: body in plugins/MauticCrmBundle/Integration/SalesforceIntegration.php on line 959
PHP Notice:  Undefined variable: body in plugins/MauticCrmBundle/Integration/SalesforceIntegration.php on line 959
PHP Notice:  Undefined variable: body in plugins/MauticCrmBundle/Integration/SalesforceIntegration.php on line 959

  [Doctrine\DBAL\Exception\SyntaxErrorException]
  An exception occurred while executing 'SELECT l.id,l.lastname, l.firstname, l.company, l.email, l.lastname, l.firstname, l.city, l.country, l
  .email FROM mau_leads l WHERE l.id not in (select i.internal_entity_id from mau_ integration_entity i where i.integration = ? and i.internal_
  entity = "lead")' with params ["Salesforce"]:
  You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'i wher
  e i.integration = ? and i.internal_entity = "lead")' at line 1
```

**Comment by mqueme** [comment]
fixed.

**Comment by escopecz** [comment]
It works now. Thanks!

**Pull request merged**